jgrapht-core/src/main/java/org/jgrapht/GraphIterables.java

```diff
@@ -0,0 +1,231 @@
+/*
+ * (C) Copyright 2020-2020, by Dimitrios Michail and Contributors.
+ *
+ * JGraphT : a free Java graph-theory library
+ *
+ * See the CONTRIBUTORS.md file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0, or the
+ * GNU Lesser General Public License v2.1 or later
+ * which is available at
+ * http://www.gnu.org/licenses/old-licenses/lgpl-2.1-standalone.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR LGPL-2.1-or-later
+ */
+package org.jgrapht;
+
+/**
+ * A collection of graph iterables suitable for graph implementations which may contain a large
+ * number of vertices and edges.
+ *
+ * @param <V> the graph vertex type
+ * @param <E> the graph edge type
+ *
+ * @author Dimitrios Michail
+ */
+public interface GraphIterables<V, E>
+{
+
+    /**
+     * Get the underlying graph.
+     * 
+     * @return the underlying graph
+     */
+    Graph<V, E> getGraph();
+
+    /**
+     * Returns an iterable over the edges of the graph.
+     * 
+     * <p>
+     * Whether the ordering is deterministic, depends on the actual graph implementation. It is the
+     * responsibility of callers who rely on this behavior to only use graph implementations which
+     * support it.
+     * 
+     * @return an iterable over the edges of the graph.
+     */
+    default Iterable<E> edges()
+    {
+        return getGraph().edgeSet();
+    }
+
+    /**
+     * Return the number of edges in the graph.
+     * 
+     * @return the number of edges.
+     */
+    default long edgeCount()
+    {
+        return getGraph().edgeSet().size();
+    }
+
+    /**
+     * Returns a set of the vertices contained in this graph. The set is backed by the graph, so
```

Comment:
For this and the other methods, the Javadoc should be in terms of iterables rather than sets.

Since we're adding a new interface, we also have an opportunity here for tightening up our specifications.  The existing Graph interface jumbles together some methods which return guaranteed live views, whereas other methods don't guarantee anything at all (and in some cases the implementation materializes the result immediately, meaning the view is definitely not live).

Since our intention is laziness here, it would be nice to be able to state uniformly that all the views defined here are live (at least with respect to read operations), meaning the result set is not determined earlier than the iterator() call.  (And subsequent iterator calls may produce different results.)  However, to do so, we would have to add a wrapper that implements the laziness (at least in cases where the Set-returning method may be eager).

What do you think?

(For writes, we can state that it's up to the implementation whether `remove` works on returned iterators.)

Reply:
Not sure I understood the "laziness" part.

Reply:
Consider allEdges(x, y).  The underlying implementation for our default graphs materializes an ArrayUnenforcedSet, which does not change subsequently if edges are added or removed between x and y.  So the returned iterable is in no way live.

There's an easy way to fix this, which is to return a wrapper implementation of Iterable.  When the wrapper's `iterator()` is invoked, we want it to call `graph.getAllEdges(x, y).iterator()`.  There's no extra cost other than one extra object (the wrapper).  And it may be possible to write a generic wrapper which holds a thunk (the invocation to be executed).

This may be overengineering since the main purposes of the iterables is one-shot use in a for loop (so liveness doesn't matter).

Reply:
Yes, this is inline with the goals of this interface!  While other methods such as `edges()` are already live, it makes sense to use such a wrapper anyway, to be independent of what the underlying graph implementation does.

Reply:
Yep.

Reply:
Just uploaded a first version